packages/framework/presence/src/presenceDatastoreManager.ts

```diff
@@ -195,14 +202,103 @@ export class PresenceDatastoreManagerImpl implements PresenceDatastoreManager {
 		return entry.public;
 	}
 
-	private localUpdate(data: DatastoreMessageContent, _forceBroadcast: boolean): void {
-		const content = {
-			sendTimestamp: Date.now(),
+	private queuedMessage: DatastoreUpdateMessage["content"] | undefined;
+
+	/**
+	 * The time at which the presence data must be sent. When presence updates are submitted, this value is calculated
+	 * based on the allowable latency for the update. A timer will fire at this time, sending any queued message
+	 * immediately. Note that if updates come in with allowable latencies that require sending before this deadline, those
+	 * messages will be sent immediately. Therefore this value can be considered the time of the next scheduled presence
+	 * update.
+	 */
+	private sendMessageDeadline: number = 0;
+
+	private localUpdate(data: DatastoreMessageContent, options: LocalUpdateOptions): void {
+		const allowableUpdateLatency = options.allowableUpdateLatency ?? 0;
+		const forceBroadcast = options.forceBroadcast;
+
+		const now = Date.now();
+		const updateDeadline = now + allowableUpdateLatency;
+		if (this.queuedMessage === undefined) {
+			// no queued message, so set the deadline based on the current message's allowable latency
+			this.sendMessageDeadline = updateDeadline;
+		}
+
+		// This function-local "datastore" will hold the merged message data
+		const queueDatastore: {
+			[WorkspaceAddress: string]: ValueElementMap<PresenceStatesSchema>;
+		} = this.queuedMessage?.data ?? {};
+
+		const currentMessageData: DatastoreMessageContent = data;
+
+		// Iterate over the current message data; individual items are workspaces
+		for (const workspaceName of Object.keys(currentMessageData)) {
+			const workspaceData = currentMessageData[workspaceName];
+
+			// Initialize the merged data as the queued datastore entry for the workspace
+			// Since the key might not exist, create an empty object in that case. It will
+			// be set explicitly after the loop.
+			const mergedData = queueDatastore?.[workspaceName] ?? {};
+
+			// Iterate over each value manager and its data, merging it as needed
+			for (const valueManagerKey of Object.keys(workspaceData)) {
+				for (const [clientSessionId, value] of brandedObjectEntries(
+					workspaceData[valueManagerKey],
+				)) {
+					// TODO: Should any values be ignored here? E.g. values with ignoreUnmonitored?
```

Comment:
For updates, each valueManager entry should have exactly one entry that is this client's session id... unless merging system workspace which is special...

See `localUpdate` on line 156. That guy does a lot of prep to form a full update message's content. Including prep of the system:presence workspace data that should be done last like sendTimestamp and averageLatency that I noted.

Probably I should have renamed the private function at some point and forceBroadcast argument should have been removed ...

The queued datastore updates should be handled in the caller.

Probably what we want is to have most of the content of line 156's localUpdate lambda be in method that sendQueuedMessage calls.
`queuedMessage` -> `queuedUpdateDatastore` (type `GeneralDatastoreMessageContent | undefined`)
From the new method remove the `updates` prep (that will be part of the merge you have now). `[internalWorkspaceAddress]: updates` will be replaced by `...queuedUpdateDatastore`.
```diff
   this.localUpdate(
    {
     // ...
     "system:presence": {
      clientToSessionId: {
       [clientConnectionId]: { ...currentClientToSessionValueState },
      },
     },
-     [internalWorkspaceAddress]: updates,
+     ...queuedUpdateDatastore,
    },
```

Then your merge is walking over more specific things: one workspace, one client session id for each value key.

That is probably far too much to handle in one read. Short is that code was not organized the best before (sorry) and the hook went in later than is ideal adding extra work and some complications. We should probably talk it through and hopefully the above text is good reference for making the changes.

Reply:
Update looks good enough for now